app/src/main/java/com/xmartlabs/template/ui/RecyclerViewGridSpacingDecoration.java

```diff
@@ -0,0 +1,253 @@
+package com.xmartlabs.template.ui;
+
+import android.graphics.Rect;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.v7.widget.GridLayoutManager;
+import android.support.v7.widget.RecyclerView;
+import android.view.View;
+
+import com.annimon.stream.Optional;
+import com.annimon.stream.Stream;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+import lombok.Builder;
+
+/**
+ * {@link RecyclerView.ItemDecoration} subclass designed to add spacing to item controlled by a {@link GridLayoutManager}.
+ *
+ * This decorator relies on both the {@link GridLayoutManager#mSpanCount} set and the
+ * {@link GridLayoutManager.SpanSizeLookup}, so both must be defined.
+ *
+ * This decorator allows setting spacing for every item, and different spacing for:
+ * <ul>
+ *   <li>First row items</li>
+ *   <li>Last row items</li>
+ *   <li>First column items</li>
+ *   <li>Last column items</li>
+ * </ul>
+ *
+ * There's another option for which you can set the spacing individually for every item, using the
+ * {@link #setItemOffsetConsumer} consumer function.
+ *
+ * Note that calculating the first and last column for each row involves some processing and can hurt performance.
+ * Because of that, those values are calculated once and then cached for faster access.
+ * If new items are added to the {@link RecyclerView}, you must invalidate the cache for the decoration to work
+ * properly, using one of the following methods:
+ * <ul>
+ *   <li>{@link #invalidateCache()} to invalidate the whole cache</li>
+ *   <li>{@link #invalidateCacheFromPosition(int)} to invalidate the cache from a given position (if you append items
+ *       to the latest position, using this will yield better performance)</li>
+ * </ul>
+ *
+ * For even faster performance, consider enabling {@link GridLayoutManager.SpanSizeLookup#setSpanIndexCacheEnabled(boolean)}.
+ */
+@Builder
+public class RecyclerViewGridSpacingDecoration extends RecyclerView.ItemDecoration {
+  private static final int FIRST_ROW_GROUP = 0;
+
+  /**
+   * Top spacing for the first row.
+   * If null, {@link #itemSpacing} will be used.
```

Comment:
In my opinion these second-sentences should go in the same paragraph as the first one.